### PR TITLE
xds: add HTTP filter modifier with additional context

### PR DESCRIPTION
**Pull request by Shikugawa**
In this PR, we introduce a mechanism to modify the HTTP filter settings given by default. The background of the need for this feature is that when Apache SkyWalking is specified as the tracing provider, the value of start_child_span in the Router Filter must be different from the default value to get the correct tracing results. See here for details. https://github.com/istio/istio/pull/32588#issuecomment-862209815

We introduce the concept of a Filter Modifier, where the context given to the Filter Modifier is changed during the HTTP Connection Manager processing, and the filter is rewritten based on the value of the context.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[x] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[x] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

File: pilot/pkg/xds/filters/modifier.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+type FilterModifierContext struct {
+	RouterFilterModifierCtx RouterFilterModifierContext
+}
+
+func ModifyFilter(ctx *FilterModifierContext, filters []*hcm.HttpFilter) {
+	for _, filter := range filters {
+		if filter.Name == wellknown.Router {
+			routerFilter := router.Router{}
+			if err := filter.GetTypedConfig().UnmarshalTo(&routerFilter); err != nil {
```
Comment (howardjohn) [inline pilot/pkg/xds/filters/modifier.go:36]: I am not sure it makes sense to double marshal in first class code. Can we just apply the filter modifier before we marshal the first time?

**Review by howardjohn — COMMENTED** [review]
cc @kyessenov this seems loosely related to some things you have been looking into possibly? Although its more first class vs arbitrary patching from user config

File: pilot/pkg/xds/filters/modifier.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+	hcm "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/network/http_connection_manager/v3"
+	"github.com/envoyproxy/go-control-plane/pkg/wellknown"
+	"istio.io/istio/pilot/pkg/networking/util"
+)
+
+type RouterFilterModifierContext struct {
+	EnableStartChildSpan bool
+}
+
+type FilterModifierContext struct {
```
Comment (adiprerepa) [inline pilot/pkg/xds/filters/modifier.go:28]: Why does `FilterModifierContext` exist when the only "subclass" is `RouterFilterModifierContext`? Why not pass in `RouterFilterModifierContext` everywhere?

File: pilot/pkg/xds/filters/modifier.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+			}
+			modifyRouterFilter(ctx.RouterFilterModifierCtx, &routerFilter)
+			filter.ConfigType = &hcm.HttpFilter_TypedConfig{
+				TypedConfig: util.MessageToAny(&routerFilter),
+			}
+		}
+	}
+}
+
+func modifyRouterFilter(ctx RouterFilterModifierContext, filter *router.Router) {
```
Comment (adiprerepa) [inline pilot/pkg/xds/filters/modifier.go:47]: why do we have this function? It has just 1 line

File: pilot/pkg/xds/modifier_test.go [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+			},
+			filtersExpected: []*hcm.HttpFilter{
+				filters.Router,
+			},
+		},
+		{
+			name: "with filter modifier ctx",
+			ctx: &filters.FilterModifierContext{
+				RouterFilterModifierCtx: filters.RouterFilterModifierContext{
+					EnableStartChildSpan: true,
```
Comment (adiprerepa) [inline pilot/pkg/xds/modifier_test.go:51]: should we have a test with `EnableStartChildSpan` set to `false`?

File: pilot/pkg/networking/core/v1alpha3/listener.go [hunk 5]
```diff
@@ -1309,7 +1309,10 @@ func buildHTTPConnectionManager(listenerOpts buildListenerOpts, httpOpts *httpLi
 
 	accessLogBuilder.setHTTPAccessLog(listenerOpts.push.Mesh, connectionManager)
 
-	configureTracing(listenerOpts, connectionManager)
+	filterModifierCtx := &xdsfilters.FilterModifierContext{}
+	configureTracing(listenerOpts, connectionManager, filterModifierCtx)
+
+	xdsfilters.ModifyFilter(filterModifierCtx, connectionManager.HttpFilters)
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/listener.go:1315]: I would suggest mutating it in the function, so can get rid of additional marshal/unmarshal

Comment (hzxuzhonghu) [reply pilot/pkg/networking/core/v1alpha3/listener.go:1315]: I mean in `configureTracing` ir even low level

File: pilot/pkg/xds/filters/modifier.go (hunk 1 above)
Comment (Shikugawa) [reply pilot/pkg/xds/filters/modifier.go]: deleted `modifier.go`. changed to build router filter in-place.

File: pilot/pkg/xds/filters/modifier.go (hunk 2 above)
Comment (Shikugawa) [reply pilot/pkg/xds/filters/modifier.go]: deleted `modifier.go`. changed to build router filter in-place.

Comment (Shikugawa) [reply pilot/pkg/xds/filters/modifier.go]: deleted `modifier.go`. changed to build router filter in-place.

File: pilot/pkg/networking/core/v1alpha3/listener.go (hunk 5 above)
Comment (Shikugawa) [reply pilot/pkg/networking/core/v1alpha3/listener.go]: deleted `modifier.go`. changed to build router filter in-place. we don't have to take care marshaling anymore.

File: pilot/pkg/networking/core/v1alpha3/tracing.go [hunk 6]
```diff
@@ -150,6 +151,8 @@ func configureFromProviderConfig(pushCtx *model.PushContext, meta *model.NodeMet
 		})
 
 	case *meshconfig.MeshConfig_ExtensionProvider_Skywalking:
+		rfCtx.StartChildSpan = true
```
Comment (howardjohn) [inline pilot/pkg/networking/core/v1alpha3/tracing.go:154]: Mutating a pointer feels tricky to understand and refactor. Would it be simpler to just return a value from configureTracing?

**Review by howardjohn — COMMENTED** [review]
/test benchmark_istio

File: pilot/pkg/networking/core/v1alpha3/tracing.go (hunk 6 above)
Comment (Shikugawa) [reply pilot/pkg/networking/core/v1alpha3/tracing.go]: fixed

**Review by howardjohn — COMMENTED** [review]
/test benchmark_istio

File: pilot/pkg/xds/filters/filters.go [hunk 7]
```diff
@@ -157,6 +151,21 @@ var (
 	HTTPMx = buildHTTPMxFilter()
 )
 
+func BuildRouterFilter(ctx *RouterFilterContext) *hcm.HttpFilter {
```
Comment (howardjohn) [inline pilot/pkg/xds/filters/filters.go:154]: Can you keep the old precompute `Router` when ctx is nil (assuming ctx is nil for cases we don't have overrides?)?

This has a moderate performance impact on allocs:
```
ListenerGeneration/gateways-8                       9.17MB ± 0%        9.55MB ± 0%   +4.13%  (p=0.008 n=5+5)
ListenerGeneration/gateways-shared-8                20.3kB ± 0%        21.1kB ± 0%   +3.70%  (p=0.008 n=5+5)
ListenerGeneration/empty-8                           827kB ± 0%         829kB ± 0%   +0.27%  (p=0.008 n=5+5)
ListenerGeneration/tls-8                             787kB ± 0%         788kB ± 0%   +0.19%  (p=0.008 n=5+5)
ListenerGeneration/telemetry-8                       942kB ± 0%         944kB ± 0%   +0.24%  (p=0.008 n=5+5)
ListenerGeneration/virtualservice-8                 80.7kB ± 0%        81.8kB ± 0%   +1.42%  (p=0.016 n=4+5)
ListenerGeneration/authorizationpolicy-8             623kB ± 0%         624kB ± 0%   +0.18%  (p=0.008 n=5+5)
ListenerGeneration/peerauthentication-8             45.4kB ± 0%        46.1kB ± 0%   +1.66%  (p=0.029 n=4+4)
ListenerGeneration/knative-gateway-8                18.8kB ± 0%        19.6kB ± 0%   +4.06%  (p=0.008 n=5+5)
```

**Review by howardjohn — COMMENTED** [review]
LGTM aside from small perf comment.

Thanks for all the improvements!

**Review by hzxuzhonghu — APPROVED** [review]

**Pull request merged**